### ISSUE TITLE
coefficients of final model

### ISSUE THREAD

**Issue by skfkrh** [microsoft/EconML#278]
I'm trying to understand how to fetch final model.
Y=theta(X)*alpha*T+f(w)
I think effect interval seems like giving the CI for coefficient alpha, but I'd like to get details of theta(X), for example, if I use featurizer to be polynomial with degree of 2, then all coefficients of theta(X)=beta1+beta2*X+beta3*X^2, then how can I get betas?

If I use final model to be LASSO, does it reduce alpha or betas? Trying to understand what's the parameters of LASSO and dependent variables. Or does it consider a model to be beta1*alpha, etc.?

**Comment by vsyrgkanis** [comment]
Check out  this customer scenario notebook and in particular cell [11]. 

The effect_interval(X) is providing inference on the quantity \theta(X)=<a, X> + intercept. You can get inference on a and intercept by calling model.summary() after fitting the model and even pass feature names. This will return inference results on the coefficients of the linear CATE model and the intercept of the CATE model.  Alternatively you can also call: coef__interval() and intercept_interval(), to get confidence intervals for these parameters, but this will not print a nice summary.

If you use a featurizer and the featurizer also has a .get_feature_names(featnames) method (same API as the polynomial featurizer of sklearn: https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.PolynomialFeatures.html#sklearn.preprocessing.PolynomialFeatures.get_feature_names ) 
Then this will even return the composite feature names, as in the example in the cell.

**Comment by skfkrh** [comment]
Thank you! 
model.summary() works when final model is linear regression and using inference to be statsmodel with LinearDMLEstimator. In case I have to use bootstrapping because final model is LASSO for DMLEstimator etc. How can I get coefficients?

When I use LASSO as final mode for multi-treatment case. Y=theta(X)(T1 + T2)+f(w), l, does the model regularize theta(X) or does it fit theta1(X) for T1 and theta2(X) for T2 so coefficients are( # features in theta) * (# treatment) for LASSO and regularize each of them as individual coefficients? Because in the documentation, I think I read there are (# outcome)*( # features in theta) * (# treatment) coefficients.

Also what's the difference between marginal_effect vs. const_marginal_effect? in what cases, will they yield different results?
I didn't see the difference between two results so far except that marginal_effect asked me to provide treatment although here it's expected that the marginal effect doesn't differ by treatment as it's just coefficients for each treatment dY/dT1, and dY/dT2 marginalizing all X and W.

**Comment by vsyrgkanis** [comment]
1. I would be cautious using bootstrap with the lasso. 
   a. The shrinkage bias introduced by the lasso will not be picked up by the bootstrap and the coverage will only capture the variance and not the bias. You might be over-confident. For correct inference with a lasso final model, I would use the "SparseLinearDMLCateEstimator" which uses the debiased lasso in the final stage to perform inference. In that case you can also call summary(), if you use "inference='debiasedlasso'".  

   b. If you want to use a Lasso final model and get the fitted coefficient out, you can use est.model_cate.coef_. This has shape (1 + n_features) * (n_treatments). The first (n_features + 1) are the params of theta_1(X), the next (n_features + 1) are of theta_2(X) etc. For each treatment, the first parameter is the intercept of the cate, i.e. a_0 in theta_i(X) = a_{i0} + <a_i, X> and the remainder are the coefficients of the features.

   c. Unfrotunately, if you still want to go with bootstrapping the lasso, I just realized that we have removed the coef_ and intercept_ from the DMLCateEstimator (most prob because we have no good handle of where the final model stores the coefficients, if the final model is not hard-coded) and these are available only for Linear and SparseLinear DML. If we had enabled these parameters (which could be accessing the corresponding parameters of the fitted lasso), then  you can get inference on the coefficients by calling: coef__interval() and intercept__interval(). 

   d. For now you can use the following ad-hoc solution and hopefully in the future we can enable coef__ and intercept__ for all DML estimators, assuming that the final model stores the coefficient in coef_ (the first entry in coef_ is the intercept of the CATE and the remainder are the coefficients in front of the features). This solution uses our legacy BootstrapEstimator wrapper and also creates a custom DML class that inherits DMLCateEstimator and exposes the coef_ of the final model cate. 
```
from econml.bootstrap import BootstrapEstimator

class CustomDML(DMLCateEstimator):
    @property
    def coef_(self):
        return self.model_cate.coef_

est2 = BootstrapEstimator(CustomDML(model_y=RandomForestRegressor(),
                        model_t=RandomForestRegressor(),
                        model_final=Lasso(alpha=0.1, fit_intercept=False),
                        featurizer=PolynomialFeatures(degree=10, include_bias=False),
                        random_state=123), n_bootstrap_samples=4)
est2.fit(Y_train, T_train, X_train, W_train)
est.coef__interval()
```

2. Yes the model runs a big lasso on the parameters that you mention. It is actuall (#features in theta + 1) * (# treatment) as each theta_i(X) also has an intercept: theta_i(X) = <a_i, X> + a_{i0}.

3. Marginal effect is different from const_marginal effect only when the model you assume is non-linear in the treatment. All DML estimators assume that the model is linear in the treatment, in which case the two functions are the same. If you had a non-linear in treatment model, i.e. y=h(X,T) + f(X, W), (e.g. as in DeepIV), then marginal effect returns tau(X, T) = nabla_T h(X, T), which is dependent on T. For such non-linear models, const_marginal_effect is not available. The const_marginal_effect is available only for estimators that inherit from `LinearCateEstimator` https://econml.azurewebsites.net/_autosummary/econml.cate_estimator.html#econml.cate_estimator.LinearCateEstimator, and all such estimators that inherit from this, assume that the model is linear in treatment, i.e. y = theta(X) * T + f(X, W). In that case const_marginal_effect is a short-cut for theta(X) and is equal to tau(X, T) for all T.

**Comment by vsyrgkanis** [comment]
Even better code. To avoid using the legacy bootstrap estimate, you can just create a custom class that adds the DebiasedLassoCateEstimatorMixin inheritance to the mix. This will make the DML cate estimate have attributes similar to the debiased lasso. Then you can use coef_, intercept_, coef__interval and intercept__interval as in the debiased lasso. These parameters are more interpretable as coef_ has shape (dy, dt, dx) and intercept has shape (dy, dt) (each dimension is omitted if either Y or T are vectors and not 2d). Thus it parses the lasso coefs into parameters in front of the features a_i and the intercept of the CATE  a_{i0}. You still wont be able to use summary(), because our bootstrap inference still has not summary() implemented, but this will be fixed in the upcoming release within a week or so.

```
from econml.inference import BootstrapInference
from econml.cate_estimator import DebiasedLassoCateEstimatorMixin

class CustomDML(DMLCateEstimator, DebiasedLassoCateEstimatorMixin):
    pass

est2 = CustomDML(model_y=RandomForestRegressor(),
                        model_t=RandomForestRegressor(),
                        model_final=Lasso(alpha=0.1, fit_intercept=False),
                        featurizer=PolynomialFeatures(degree=10, include_bias=False),
                        random_state=123)
est2.fit(Y_train, T_train, X_train, W_train,
         inference=BootstrapInference(n_bootstrap_samples=4))
```

**Comment by skfkrh** [comment]
Thank you!

**Comment by skfkrh** [comment]
Just a qq: you mentioned that using Lasso in the final model will lead to incorrect inference. Will it happen the same, if I use Lasso as final model for IntentToTreatDRIV and DMLIV?

**Comment by vsyrgkanis** [comment]
Yes that's the same issue in that too. You can use debiased lasso instead of lasso again here too. Unfortunately `SparseLinearIntentToTreatDRIV` is not implemented yet, but below is a barebones implementation of it that you can use for now (similar in spirit to the previous code that adds the mixin). If you use a flexible final model such as a lasso with many features, then I you could use the opt_reweighted=True, that "assumes the CATE model is well-specified and reduces variance" and set covariance clipping to almost zero to reduce variance of estimates. 
```python
from econml.sklearn_extensions.linear_model import MultiOutputDebiasedLasso, WeightedLassoCV
from econml.ortho_iv import IntentToTreatDRIV
from econml.cate_estimator import DebiasedLassoCateEstimatorMixin
from econml.inference import LinearModelFinalInference

class SparseLinearIntentToTreatDRIV(DebiasedLassoCateEstimatorMixin, IntentToTreatDRIV): 
    pass

est = SparseLinearIntentToTreatDRIV(model_Y_X=RandomForestRegressor(min_samples_leaf=20),
                                    model_T_XZ=RandomForestClassifier(min_samples_leaf=20),
                                    flexible_model_effect=WeightedLassoCV(),
                                    final_model_effect=MultiOutputDebiasedLasso(fit_intercept=False),
                                    n_splits=5, cov_clip=1e-6, opt_reweighted=False)

n = 6000
d = 10
X = np.random.normal(size=(n, d))
Z = np.random.binomial(1, .5, size=(n,))
v = np.random.binomial(1, .5, size=(n,))
T = np.random.binomial(1, .5 + .3*(2*Z-1) + .1*(2*v-1))
y = X[:, 0]*T + v + np.random.normal(0, .1, size=(n,))

est.fit(y, T, Z, X=X, inference=LinearModelFinalInference())
est.summary()
```

![image](https://user-images.githubusercontent.com/13246065/92580598-917ca700-f25c-11ea-9830-6c446c445cd1.png)

**Comment by skfkrh** [comment]
Thank you!

Just another qq: what's the difference between flexible_model_effect and final_model_effect? 
Seems like when I don't indicate final_model_effect, it's using flexible_model_effect anyway. What's the case of having them different? Is there any explanation of it in the paper?

By the way, not sure if it's a bug but wanted to leave a note, nothing urgent; just what I found.
I played with multiple IV approaches here with the same data sets, but it seems like IntentToTreatDRIV model with the option of inference="bootstrap" is giving an error. 
when I execute the code _model.fit(Y.values, T, Z, X_data.values, inference="bootstrap")_
I got the following error.

[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 8 concurrent workers.
Traceback (most recent call last):

  File "<ipython-input-1006-c65f205cb875>", line 1, in <module>
    modelnon2.fit(Y.values, T, Z, X_data.values, inference="bootstrap")

  File "C:\Users\minkan\AppData\Local\Continuum\anaconda3\lib\site-packages\econml\ortho_iv.py", line 914, in fit
    sample_weight=sample_weight, sample_var=sample_var, inference=inference)

  File "C:\Users\minkan\AppData\Local\Continuum\anaconda3\lib\site-packages\econml\cate_estimator.py", line 92, in call
    inference.fit(self, Y, T, *args, **kwargs)

  File "C:\Users\minkan\AppData\Local\Continuum\anaconda3\lib\site-packages\econml\inference.py", line 56, in fit
    est.fit(*args, **kwargs)

  File "C:\Users\minkan\AppData\Local\Continuum\anaconda3\lib\site-packages\econml\bootstrap.py", line 77, in fit
    for obj, inds in zip(self._instances, indices)

  File "C:\Users\minkan\AppData\Local\Continuum\anaconda3\lib\site-packages\joblib\parallel.py", line 1042, in __call__
    self.retrieve()

  File "C:\Users\minkan\AppData\Local\Continuum\anaconda3\lib\site-packages\joblib\parallel.py", line 921, in retrieve
    self._output.extend(job.get(timeout=self.timeout))

  File "C:\Users\minkan\AppData\Local\Continuum\anaconda3\lib\multiprocessing\pool.py", line 657, in get
    raise self._value

  File "C:\Users\minkan\AppData\Local\Continuum\anaconda3\lib\multiprocessing\pool.py", line 121, in worker
    result = (True, func(*args, **kwds))

  File "C:\Users\minkan\AppData\Local\Continuum\anaconda3\lib\site-packages\joblib\_parallel_backends.py", line 595, in __call__
    return self.func(*args, **kwargs)

  File "C:\Users\minkan\AppData\Local\Continuum\anaconda3\lib\site-packages\joblib\parallel.py", line 253, in __call__
    for func, args, kwargs in self.items]

  File "C:\Users\minkan\AppData\Local\Continuum\anaconda3\lib\site-packages\joblib\parallel.py", line 253, in <listcomp>
    for func, args, kwargs in self.items]

  File "C:\Users\minkan\AppData\Local\Continuum\anaconda3\lib\site-packages\econml\bootstrap.py", line 68, in fit
    x.fit(*args, **kwargs)

TypeError: fit() got an unexpected keyword argument 'W'

**Comment by vsyrgkanis** [comment]
The flexible_model_effect is the model that is used to estimate the preliminary estimate of the CATE, which is treated as a nuisance function and is then "debiased" in the final step. It is the model that estimates ``\theta_{pre}` in Section 5 of the paper:
https://arxiv.org/pdf/1905.10176.pdf

The final_model_effect is the model that  is estimated in the final step. 

One point is that the final model does not really need to contain the true CATE model in its function space, so it can be as simple as a "constant model" which will the return the ATE. Or a linear model that only uses a specific variable in X, which will estimate the projection of the CATE on that space of functions. However, the preliminary model needs to be asymptotically consistent for the true CATE. So that model should be more flexible, i.e. you can use a LassoCV with some featurizer, or a Random or Gradient boosted forest (albeit I would regularize quite a lot to reduce variance). 

Also if you want confidence intervals then your final model needs to support confidence intervals, e.g. a `MultiOutputDebiasedLasso` or a `StatsModelsLinearRegression` or a https://github.com/microsoft/EconML/blob/ffdfda8bf733313fde4f5f762b6fa3c62d7e1531/econml/sklearn_extensions/ensemble.py#L97

Your preliminary flexible model effect need not support confidence intervals.

**Comment by vsyrgkanis** [comment]
@kbattocchi I'm not sure what the reason for the final error is with the bootstrap?

**Comment by skfkrh** [comment]
Thank you!

**Comment by vsyrgkanis** [comment]
The coef and intercept is addressed now by #283 and would be there in the next release.

**Comment by xjtusjtu** [comment]
> 1. I would be cautious using bootstrap with the lasso.
>    a. The shrinkage bias introduced by the lasso will not be picked up by the bootstrap and the coverage will only capture the variance and not the bias. You might be over-confident. For correct inference with a lasso final model, I would use the "SparseLinearDMLCateEstimator" which uses the debiased lasso in the final stage to perform inference. In that case you can also call summary(), if you use "inference='debiasedlasso'".
>    b. If you want to use a Lasso final model and get the fitted coefficient out, you can use est.model_cate.coef_. This has shape (1 + n_features) * (n_treatments). The first (n_features + 1) are the params of theta_1(X), the next (n_features + 1) are of theta_2(X) etc. For each treatment, the first parameter is the intercept of the cate, i.e. a_0 in theta_i(X) = a_{i0} + <a_i, X> and the remainder are the coefficients of the features.
>    c. Unfrotunately, if you still want to go with bootstrapping the lasso, I just realized that we have removed the coef_ and intercept_ from the DMLCateEstimator (most prob because we have no good handle of where the final model stores the coefficients, if the final model is not hard-coded) and these are available only for Linear and SparseLinear DML. If we had enabled these parameters (which could be accessing the corresponding parameters of the fitted lasso), then  you can get inference on the coefficients by calling: coef__interval() and intercept__interval().
>    d. For now you can use the following ad-hoc solution and hopefully in the future we can enable coef__ and intercept__ for all DML estimators, assuming that the final model stores the coefficient in coef_ (the first entry in coef_ is the intercept of the CATE and the remainder are the coefficients in front of the features). This solution uses our legacy BootstrapEstimator wrapper and also creates a custom DML class that inherits DMLCateEstimator and exposes the coef_ of the final model cate.
> 
> from econml.bootstrap import BootstrapEstimator
> 
> class CustomDML(DMLCateEstimator):
>     @property
>     def coef_(self):
>         return self.model_cate.coef_
> 
> est2 = BootstrapEstimator(CustomDML(model_y=RandomForestRegressor(),
>                         model_t=RandomForestRegressor(),
>                         model_final=Lasso(alpha=0.1, fit_intercept=False),
>                         featurizer=PolynomialFeatures(degree=10, include_bias=False),
>                         random_state=123), n_bootstrap_samples=4)
> est2.fit(Y_train, T_train, X_train, W_train)
> est.coef__interval()
> 2. Yes the model runs a big lasso on the parameters that you mention. It is actuall (#features in theta + 1) * (# treatment) as each theta_i(X) also has an intercept: theta_i(X) = <a_i, X> + a_{i0}.
> 3. Marginal effect is different from const_marginal effect only when the model you assume is non-linear in the treatment. All DML estimators assume that the model is linear in the treatment, in which case the two functions are the same. If you had a non-linear in treatment model, i.e. y=h(X,T) + f(X, W), (e.g. as in DeepIV), then marginal effect returns tau(X, T) = nabla_T h(X, T), which is dependent on T. For such non-linear models, const_marginal_effect is not available. The const_marginal_effect is available only for estimators that inherit from `LinearCateEstimator` https://econml.azurewebsites.net/_autosummary/econml.cate_estimator.html#econml.cate_estimator.LinearCateEstimator, and all such estimators that inherit from this, assume that the model is linear in treatment, i.e. y = theta(X) * T + f(X, W). In that case const_marginal_effect is a short-cut for theta(X) and is equal to tau(X, T) for all T.

_Comment_: I think all the $\tau(X,T)$ above are actually $\partial \tau(x,T)$. There is a difference in the meaning of these two notations.
